lib/fabrik/VirtualHostManager.js

```diff
@@ -0,0 +1,89 @@
+const _ = require('lodash');
+const Promise = require('bluebird');
+const bosh = require('../bosh');
+const utils = require('../utils');
+const VirtualHostInstance = require('./VirtualHostInstance');
+const BaseManager = require('./BaseManager');
+const VirtualHostAgent = require('./VirtualHostAgent');
+const logger = require('../logger');
+const errors = require('../errors');
+const BadRequest = errors.BadRequest;
+const ServiceBindingAlreadyExists = errors.ServiceBindingAlreadyExists;
+const ServiceBindingNotFound = errors.ServiceBindingNotFound;
+const ServiceInstanceNotFound = errors.ServiceInstanceNotFound;
+const NotFound = errors.NotFound;
+const mapper = require('./VirtualHostRelationMapper');
+
+class VirtualHostManager extends BaseManager {
+  constructor(plan) {
+    super(plan);
+    this.director = bosh.director;
+    this.agent = new VirtualHostAgent(this.settings.agent);
+    this.mapper = mapper.VirtualHostRelationMapper;
+  }
+
+  createVirtualHost(deploymentName, params, instanceId) {
+    logger.info(`Creating virtual host for deployment '${deploymentName}'...`);
+    return this.director.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.createVirtualHost(ips, instanceId))
+      .then(() => this.mapper.createVirtualHostRelation(deploymentName, instanceId))
+      .tap(() => logger.info('+-> Created virtual host'));
+  }
+
+  deleteVirtualHost(deploymentName, instanceId) {
+    logger.info(`Deleting virtual host for deployment '${deploymentName}'...`);
+    return this.director.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.deleteVirtualHost(ips, instanceId))
+      .then(() => this.mapper.deleteVirtualHostRelation(instanceId))
```

Comment:
Delete involves deleting the vhost and deleting the mapping info on container. There can be a scenario where delete of vhost may succeed but delete vhost mapping might fail, which would result in an error and instance deletion will fail. 
now on retry, in the above flow delete virtual host would most probably throw 404 & we wont even attempt to delete vhost relation mapping. When delete vhost says 404, we must handle it and still still go ahead and try to delete this vhost relation.
This can be done in couple of ways, one of the below is my suggestion:
```
return this.director.getDeploymentIps(deploymentName)
           .then(ips => Promise.all([this.agent.deleteVirtualHost(ips, instanceId),
                                                     this.mapper.deleteVirtualHostRelation(instanceId)]);
```
Now doing so ensures both of these happen every time. Ofcse, if deployment itself is deleted `this.director.getDeploymentIps` would throw 404, but i guess there would be some checks not to delete deployment when there are vhosts on it right?

Reply:
I will incorporate above changes.

For deletion of deployment:
I implemented in agent /deprovision call while delete deployment not to allow deletion if some virtual hosts already exists.
But, It also has an issue that in DirectorInstance, deleteSecurityGroup happens in initialize() without even waiting for /deprovision called in manager.deleteDeployment to respond. So even if /deprovision fails, service-instance becomes non operational becuase of deletion of security-group.
Moving security group deletion from https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/lib/fabrik/DirectorInstance.js#L56 to https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/lib/fabrik/DirectorInstance.js#L170 should solve it.
What do you suggest?

Reply:
One other way is:
return this.director.getDeploymentIps(deploymentName)
      .then(ips => this.agent.deleteVirtualHost(ips, instanceId))
      .then(() => this.mapper.deleteVirtualHostRelation(instanceId))
      .tap(() => logger.info(`+-> Deleted virtual host for deployment '${deploymentName}'`))
      .catch(NotFound, err => {
        /* jshint unused:false */
        this.mapper.deleteVirtualHostRelation(instanceId);
        throw new ServiceInstanceNotFound(instanceId);
      });

In case of any NotFound issue it will perform clean up task of removing entry from the provider.